### ISSUE TITLE
How to close a connection?

### ISSUE THREAD

**Issue by liron-navon** [go-resty/resty#151]
How do I close the connection like in net/http ?

```
res, err := http.Get(url)

res.Body.Close()
```

**Comment by jeevatkm** [comment]
@liron-navon by default `resty` take cares of it automatically except for [`SetDoNotParseResponse`](https://godoc.org/github.com/go-resty/resty#Client.SetDoNotParseResponse) scenario, its caller responsibility to close. This option is available at Client and each Request level.

**Comment by jeevatkm** [comment]
@liron-navon does it satisfy your need?

**Comment by jeevatkm** [comment]
@liron-navon Closing it, if you have any queries open up a new one.

**Comment by MidnightSong** [comment]
@jeevatkm thank you for your answer,I'm trying to find the solution to this problem

**Comment by SeanTolstoyevski** [comment]
When we create a default client with `New` in the v2 version, are `Close` operations performed automatically for responses?

We have a project where we use Resty heavily and there seems to be a memory leak. Resty is also among the suspects :) .

**Comment by jeevatkm** [comment]
@SeanTolstoyevski Behavior noted in this comment - https://github.com/go-resty/resty/issues/151#issuecomment-387501756

**Comment by jeevatkm** [comment]
@mortymacs Typically, the underlying Go HTTP client uses connection pooling by default to use a resue connection with a particular Host. It closes the connection the after `IdleConnTimeout`.  Also read this https://github.com/go-resty/resty/issues/151#issuecomment-387501756.

You can do the same with Resty, but it is not recommended for a production use case. 
```go
restyClient := resty.New().R()
...
defer restyClient.GetClient().CloseIdleConnections()
...
```

What is the goal you're trying to achieve?

**Comment by jeevatkm** [comment]
@mortymacs Also, you can use the Resty provided method; but it is not recommended to close the connection every time due to performance.
https://pkg.go.dev/github.com/go-resty/resty/v2#Client.SetCloseConnection

**Comment by mortymacs** [comment]
Thanks for the solution!
Sometimes I need to download something or send a request to an API and then step away. So, I need to close the connection until I need it again. I understand your concerns about performance, which are definitely important. For many of my services that need ongoing connections to various upstreams, it makes sense to leave the connection open rather than closing it each time.
Thanks, @jeevatkm !

**Comment by winterant** [comment]
If I want to use connection pool, how to do?
In other word, a tcp connection don't close and next request reuse it. But i tested resty v3 i found that resty would hold a connection although response is done.
I tried setTimeout(), it will release connection after timeout although response already ok.

**Comment by kkHAIKE** [comment]
In Resty v3, even without setting `SetDoNotParseResponse`, the response still needs to be manually closed by default (if neither `SetResponseBodyUnlimitedReads` nor `SetDebug` is set). This change is not documented, which led to a file descriptor exhaustion incident in production.

**Comment by sebjung** [comment]
EDIT: Sorry for the noise, we found our problem. We forgot to call `SetResponse` and received a body in the response, which was never parsed and therefore never closed. 

Original message:
@jeevatkm We just ran into a similar issue as @kkHAIKE mentioned. We haven't set `SetDoNotParseResponse` and we're not accessing the body directly, we're only using SetError, yet still we none of our connections get closed. You mentioned to not invoke `Client.SetCloseConnection`. What would be the "proper" way to solve this issue from your perspective?
For now it looks like having a `defer resp.RawResponse.Body.Close()` or `resp.Body.Close()` solved the issue for us. But is there a difference between the two bodies?
